[Scala 3] Add support for given imports

tgodzik opened the pull request:
Previously, organize-imports-rule would throw an exception in case of given imports, now it should sort them correctly. Currently they will be sorted separately at the end, but we can add some more options. However I am not sure what will be the best practises here. We could possibly have an option to put them at the end, but I think the current situation is fine for now.

~~You cannot unimport given imports, so we don't have do do anything similar to wildcard improts when it comes to given wildcard imports.~~ You can actually

~~Tests are currently blocked on https://github.com/liancheng/scalafix-organize-imports/pull/179 , I only checked it manually.~~ We have tests now! :tada: 

@bjaglin should we wait for that?

I thought of raising an issue, but turned out faster to fix it :sweat_smile: If there are some other plans here, feel free to close the PR

![dotty-example-project-1623409805016](https://user-images.githubusercontent.com/3807253/121677603-63134d00-cab6-11eb-83b1-898a948e11ef.gif)

liancheng commented:
@tgodzik, thanks for this! Could you please also add some test cases?

bjaglin commented:
> @tgodzik, thanks for this! Could you please also add some test cases?

It's blocked by https://github.com/liancheng/scalafix-organize-imports/pull/179. I didn't have time to update the `README` to mention the fact that `removeUnused` is not supported on Scala3, but I think this could be done separately. So I would suggest we merge https://github.com/liancheng/scalafix-organize-imports/pull/179 to unlock scala3-specific tests and iterate in other PRs.

liancheng commented on rules/src/main/scala/fix/OrganizeImports.scala:
```diff
@@ -774,40 +777,55 @@ object OrganizeImports {
         List[Importee.Rename],
         List[Importee.Unimport],
+        List[Importee.Given],
+        Option[Importee.GivenAll],
```
Would you please also update the comment on this method to reflect this change?

tgodzik replied:
Added!

liancheng commented on rules/src/main/scala/fix/OrganizeImports.scala:
```diff
@@ -510,9 +511,9 @@ class OrganizeImports(config: OrganizeImportsConfig) extends SemanticRule("Organ
     config.coalesceToWildcardImportThreshold
-      .filter(names.length > _)
+      .filter(names.length + givens.length > _)
       .map(_ => importer.copy(importees = renames ++ unimports :+ Importee.Wildcard()))
```
Should the optional `Importee.GivenAll` instance also exist in the result importer?

tgodzik replied:
I handle it now separately with 4 different cases.

liancheng commented:
I'm not familiar with Scala 3 `given` imports (just learned it by reading the spec while reviewing this PR), but I wonder do we also need to update the `mergeImporters` method? It is the most complicated method of this rule and handles a bunch of corner cases.

For example, what should be the result import if we merged the following two?

```scala
import foo.{aGiven as _, given}
import foo.given
```

TBH, I'm not even sure whether the above two imports are valid, but my intention is:

- `foo.aGiven` is a given instance
- The first import intends to import all given instances under `foo` except for `aGiven` (assuming that we can unimport a given instance)
- The second import intends to import all given instances under `foo`

I think the result should be `import foo.given`.

This example is derived from the following case in Scala 2:

```scala
import foo.{x => _, _}
import foo._
```

The above Scala 2 imports should be merged into a single `import foo._`, because the unimport is canceled by the second wildcard import.

We may want to review all the cases handled in `mergeImporters` to handle all the potentially introduced new cases.

tgodzik commented:
So from what I understand there is no way to unimport given instances, however I am not 100% sure. I will check out the situation when explicitly importing a given instead of by type

tgodzik commented:
I just checked and it seems that we can indeed unimport givens, I will fix it and add some tests as soon as https://github.com/liancheng/scalafix-organize-imports/pull/179 is merged. Or should we merge it beforehand in order to fix the Scala 3 support?

tgodzik commented:
@liancheng I added a bunch of tests and added support for given inside `mergeImporters` method. It should now handle wildcard given import them the same as a simple wildcard one. There is a single difference, if we merge imports we will not merge given and normal imports, since we don't know if given imports aren't using the types from the same package. And those given imports will break if the type they are using is imported in the same place.

I think it works correctly now, but my brain is having some issues figuring all the possible edge cases :sweat_smile: 

I haven't added any support for unused imports yet since there is no option for it in the Scala 3 compiler yet, so I am not sure even if unused given imports will be reported.

Let me know what you think!

tgodzik commented on output/src/main/scala-3/fix/GroupedGivenImportsMergeUnimports.scala:
```diff
@@ -0,0 +1,8 @@
+
+import fix.GivenImports._
+import fix.GivenImports.{gamma => _, given Beta, given Zeta, given}
+import fix.GivenImports2.{alpha => _, beta => _}
```
As you can see the imports are separated into givens and normal ones aside from the situation, where the unimports were together with given imports. Not sure if we can do better. Maybe we could check if the the types of the givens are not imported in the previous imports and join them in that case?

bjaglin reviewed (commented):
Cool! The actual implementation goes beyond my knowledge, but just as a note: you might want to remove/update https://github.com/liancheng/scalafix-organize-imports/blame/master/README.adoc#L92 since test coverage seems quite decent.

liancheng commented:
@tgodzik, thanks for the update! The `mergeImporters` is indeed tricky. If I had learned that organizing Scala imports is this complicated, I possibly wouldn't even start this project 😅

For unused imports for Scala 3, it's OK to leave it out for now as long as:

1. We document this clearly in the README.
2. No regression for existing Scala 2 support (I believe the current test coverage is good enough for this).

I will find some time to do a more thorough review, hopefully later this week.

liancheng commented:
@tgodzik, we also need to update the `ImporterExtension.isCurlyBraced` method to include `given` imports. This may affect importer sorting order.

tgodzik commented:
> @tgodzik, we also need to update the `ImporterExtension.isCurlyBraced` method to include `given` imports. This may affect importer sorting order.

Sure, I will take a look next week!

tgodzik commented:
> isCurlyBraced

It seems that we can actually skip `{}` in given imports, so the behaviour is the same as Importee.Name, but I need to fix it in Scalameta and we can improve it further on.

I also added a bit in the readme.

liancheng commented:
I'm merging this. There are a few minor places that can be cleaned up a little bit, but I can do it in a follow-up PR. Thanks again, @tgodzik!

tgodzik commented:
Thanks a lot!
